File: src/common/terminal.zig
const std = @import("std");
const io = @import("io.zig");

pub const TerminalError = error{
    NotATerminal,
    IoError,
    OutOfMemory,
};

/// Retrieves the system's window size in Zig using the standard library.
fn getWinsize() ?std.posix.winsize {
    const stdout = std.fs.File.stdout();
    if (!std.posix.isatty(stdout.handle)) return null;
    var wsz: std.posix.winsize = undefined;
    const err = std.posix.system.ioctl(stdout.handle, std.posix.T.IOCGWINSZ, @intFromPtr(&wsz));
    if (std.posix.errno(err) == .SUCCESS) return wsz;
    return null;
}

/// Returns the terminal width in bytes for the current screen size.
pub fn getTerminalWidth() usize {
    if (getWinsize()) |size| return @intCast(size.col);
    return 80;
}

/// Returns the terminal height in bytes for the current Zig source file.
pub fn getTerminalHeight() usize {
    if (getWinsize()) |size| return @intCast(size.row);
    return 24;
}

/// Checks if the provided string is a valid terminal string, returning true or false.
pub fn isTerminal() bool {
    const stdout = std.fs.File.stdout();
    return std.posix.isatty(stdout.handle);
}

/// Validates a boolean question input, returning true or false based on the provided value.
pub fn confirm(question: []const u8, default: bool) !bool {
    var ws: io.WriterState = .{};
    ws.initStdout();
    const stdout = ws.writer();
    var rs: io.ReaderState = .{};
    rs.initStdin();
    const stdin = rs.reader();

    if (default) {
        try stdout.print("{s} [Y/n]: ", .{question});
    } else {
        try stdout.print("{s} [y/N]: ", .{question});
    }
    try stdout.flush();

    if (stdin.takeDelimiterInclusive('\n') catch null) |line| {
        const trimmed = std.mem.trim(u8, line, " \t\r\n");
        if (trimmed.len == 0) return default;
        if (std.ascii.eqlIgnoreCase(trimmed, "y") or std.ascii.eqlIgnoreCase(trimmed, "yes")) return true;
        if (std.ascii.eqlIgnoreCase(trimmed, "n") or std.ascii.eqlIgnoreCase(trimmed, "no")) return false;
        return default;
    }
    return default;
}

/// Retrieves a response from an allocator using a Zig array and question parameters.
pub fn ask(allocator: std.mem.Allocator, question: []const u8, default: []const u8) ![]const u8 {
    var ws: io.WriterState = .{};
    ws.initStdout();
    const stdout = ws.writer();
    var rs: io.ReaderState = .{};
    rs.initStdin();
    const stdin = rs.reader();

    if (default.len > 0) {
        try stdout.print("{s} [{s}]: ", .{ question, default });
    } else {
        try stdout.print("{s}: ", .{question});
    }
    try stdout.flush();

    if (stdin.takeDelimiterInclusive('\n') catch null) |line| {
        const trimmed = std.mem.trim(u8, line, " \t\r\n");
        if (trimmed.len == 0) {
            return try allocator.dupe(u8, default);
        }
        return try allocator.dupe(u8, trimmed);
    }
    return try allocator.dupe(u8, default);
}

/// Retrieves an integer value from an allocator using a Zig array as input.
pub fn askInt(allocator: std.mem.Allocator, question: []const u8, default: ?i64) !i64 {
    _ = allocator;
    var ws: io.WriterState = .{};
    ws.initStdout();
    const stdout = ws.writer();
    var rs: io.ReaderState = .{};
    rs.initStdin();
    const stdin = rs.reader();

    if (default) |d| {
        try stdout.print("{s} [{d}]: ", .{ question, d });
    } else {
        try stdout.print("{s}: ", .{question});
    }
    try stdout.flush();

    if (stdin.takeDelimiterInclusive('\n') catch null) |line| {
        const trimmed = std.mem.trim(u8, line, " \t\r\n");
        if (trimmed.len == 0) {
            if (default) |d| return d;
            return error.InvalidInteger;
        }
        return std.fmt.parseInt(i64, trimmed, 10) catch error.InvalidInteger;
    }
    if (default) |d| return d;
    return error.InvalidInteger;
}

/// Manages progress tracking UI; owned by the application; ensures consistent state across sessions.
pub const ProgressBar = struct {
    description: []const u8,
    current: usize,
    total: usize,
    width: usize,

    pub fn init(description: []const u8, total: usize) ProgressBar {
        return .{
            .description = description,
            .current = 0,
            .total = total,
            .width = getTerminalWidth(),
        };
    }

    pub fn advance(self: *ProgressBar, amount: usize) void {
        self.current += amount;
        if (self.current > self.total) {
            self.current = self.total;
        }
    }

    pub fn set(self: *ProgressBar, value: usize) void {
        self.current = if (value > self.total) self.total else value;
    }

    pub fn render(self: *const ProgressBar, writer: anytype) !void {
        const bar_width = @min(self.width - self.description.len - 20, @as(usize, 40));
        const percent = if (self.total > 0) @as(usize, (@as(u64, self.current) * 100) / @as(u64, self.total)) else 0;
        const filled = if (self.total > 0) (self.current * bar_width) / self.total else 0;

        try writer.writeAll("\r");
        try writer.print("{s} [", .{self.description});

        var i: usize = 0;
        while (i < bar_width) : (i += 1) {
            if (i < filled) {
                try writer.writeAll("=");
            } else {
                try writer.writeAll(" ");
            }
        }

        try writer.print("] {d:>3}% ({d}/{d})", .{ percent, self.current, self.total });
    }

    pub fn complete(self: *const ProgressBar, writer: anytype) !void {
        try self.render(writer);
        try writer.writeAll("\n");
    }
};

/// Defines a color enum for terminal UI; managed centrally with ownership model; immutable values ensure consistency.
pub const Color = enum {
    reset,
    black,
    red,
    green,
    yellow,
    blue,
    magenta,
    cyan,
    white,
    bright_black,
    bright_red,
    bright_green,
    bright_yellow,
    bright_blue,
    bright_magenta,
    bright_cyan,
    bright_white,

    pub fn code(self: Color) []const u8 {
        return switch (self) {
            .reset => "\x1b[0m",
            .black => "\x1b[30m",
            .red => "\x1b[31m",
            .green => "\x1b[32m",
            .yellow => "\x1b[33m",
            .blue => "\x1b[34m",
            .magenta => "\x1b[35m",
            .cyan => "\x1b[36m",
            .white => "\x1b[37m",
            .bright_black => "\x1b[90m",
            .bright_red => "\x1b[91m",
            .bright_green => "\x1b[92m",
            .bright_yellow => "\x1b[93m",
            .bright_blue => "\x1b[94m",
            .bright_magenta => "\x1b[95m",
            .bright_cyan => "\x1b[96m",
            .bright_white => "\x1b[97m",
        };
    }
};

/// Prints a colored string using the provided writer, color, and text data.
pub fn colorPrint(writer: anytype, color: Color, text: []const u8) !void {
    try writer.writeAll(color.code());
    try writer.writeAll(text);
    try writer.writeAll(Color.reset.code());
}

const testing = std.testing;

test "getTerminalWidth returns at least 80" {
    const width = getTerminalWidth();
    try testing.expect(width >= 80);
}

test "getTerminalHeight returns at least 24" {
    const height = getTerminalHeight();
    try testing.expect(height >= 24);
}

test "Color codes are valid ANSI sequences" {
    try testing.expectEqualStrings("\x1b[0m", Color.reset.code());
    try testing.expectEqualStrings("\x1b[31m", Color.red.code());
    try testing.expectEqualStrings("\x1b[32m", Color.green.code());
}

test "ProgressBar init and advance" {
    var bar = ProgressBar.init("Testing", 100);
    try testing.expectEqual(@as(usize, 0), bar.current);
    try testing.expectEqual(@as(usize, 100), bar.total);

    bar.advance(25);
    try testing.expectEqual(@as(usize, 25), bar.current);

    bar.set(50);
    try testing.expectEqual(@as(usize, 50), bar.current);
}

File: src/common/src/terminal.rs
use std::io::{self, BufRead, IsTerminal, Write};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Color {
    Reset,
    Black,
    Red,
    Green,
    Yellow,
    Blue,
    Magenta,
    Cyan,
    White,
    BrightBlack,
    BrightRed,
    BrightGreen,
    BrightYellow,
    BrightBlue,
    BrightMagenta,
    BrightCyan,
    BrightWhite,
}

impl Color {
    pub fn code(self) -> &'static str {
        match self {
            Color::Reset => "\x1b[0m",
            Color::Black => "\x1b[30m",
            Color::Red => "\x1b[31m",
            Color::Green => "\x1b[32m",
            Color::Yellow => "\x1b[33m",
            Color::Blue => "\x1b[34m",
            Color::Magenta => "\x1b[35m",
            Color::Cyan => "\x1b[36m",
            Color::White => "\x1b[37m",
            Color::BrightBlack => "\x1b[90m",
            Color::BrightRed => "\x1b[91m",
            Color::BrightGreen => "\x1b[92m",
            Color::BrightYellow => "\x1b[93m",
            Color::BrightBlue => "\x1b[94m",
            Color::BrightMagenta => "\x1b[95m",
            Color::BrightCyan => "\x1b[96m",
            Color::BrightWhite => "\x1b[97m",
        }
    }
}

pub fn get_terminal_width() -> usize {
    #[cfg(unix)]
    {
        use std::mem::MaybeUninit;
        use std::os::fd::AsRawFd;
        let mut ws: MaybeUninit<libc::winsize> = MaybeUninit::uninit();
        let ret = unsafe {
            libc::ioctl(io::stdout().as_raw_fd(), libc::TIOCGWINSZ, ws.as_mut_ptr())
        };
        if ret == 0 {
            let ws = unsafe { ws.assume_init() };
            if ws.ws_col > 0 {
                return ws.ws_col as usize;
            }
        }
    }
    80
}

pub fn get_terminal_height() -> usize {
    #[cfg(unix)]
    {
        use std::mem::MaybeUninit;
        use std::os::fd::AsRawFd;
        let mut ws: MaybeUninit<libc::winsize> = MaybeUninit::uninit();
        let ret = unsafe {
            libc::ioctl(io::stdout().as_raw_fd(), libc::TIOCGWINSZ, ws.as_mut_ptr())
        };
        if ret == 0 {
            let ws = unsafe { ws.assume_init() };
            if ws.ws_row > 0 {
                return ws.ws_row as usize;
            }
        }
    }
    24
}

pub fn is_terminal() -> bool {
    io::stdout().is_terminal()
}

pub fn confirm(question: &str, default: bool) -> io::Result<bool> {
    confirm_with(question, default, &mut io::stdin().lock(), &mut io::stdout())
}

pub fn confirm_with(
    question: &str,
    default: bool,
    reader: &mut impl BufRead,
    writer: &mut impl Write,
) -> io::Result<bool> {
    let prompt = if default {
        format!("{question} [Y/n]: ")
    } else {
        format!("{question} [y/N]: ")
    };
    write!(writer, "{prompt}")?;
    writer.flush()?;
    let mut input = String::new();
    reader.read_line(&mut input)?;
    let trimmed = input.trim().to_lowercase();
    match trimmed.as_str() {
        "y" | "yes" => Ok(true),
        "n" | "no" => Ok(false),
        _ => Ok(default),
    }
}

pub fn ask(question: &str, default: &str) -> io::Result<String> {
    ask_with(question, default, &mut io::stdin().lock(), &mut io::stdout())
}

pub fn ask_with(
    question: &str,
    default: &str,
    reader: &mut impl BufRead,
    writer: &mut impl Write,
) -> io::Result<String> {
    let prompt = if default.is_empty() {
        format!("{question}: ")
    } else {
        format!("{question} [{default}]: ")
    };
    write!(writer, "{prompt}")?;
    writer.flush()?;
    let mut input = String::new();
    reader.read_line(&mut input)?;
    let trimmed = input.trim();
    if trimmed.is_empty() {
        Ok(default.to_string())
    } else {
        Ok(trimmed.to_string())
    }
}

pub fn ask_int(question: &str, default: Option<i64>) -> io::Result<i64> {
    ask_int_with(question, default, &mut io::stdin().lock(), &mut io::stdout())
}

pub fn ask_int_with(
    question: &str,
    default: Option<i64>,
    reader: &mut impl BufRead,
    writer: &mut impl Write,
) -> io::Result<i64> {
    let prompt = if let Some(d) = default {
        format!("{question} [{d}]: ")
    } else {
        format!("{question}: ")
    };
    write!(writer, "{prompt}")?;
    writer.flush()?;
    let mut input = String::new();
    reader.read_line(&mut input)?;
    let trimmed = input.trim();
    if trimmed.is_empty() {
        default.ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "no input and no default"))
    } else {
        trimmed.parse::<i64>().map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid integer"))
    }
}

#[derive(Debug)]
pub struct ProgressBar {
    description: String,
    current: usize,
    total: usize,
    width: usize,
}

impl ProgressBar {
    pub fn new(description: &str, total: usize) -> Self {
        Self {
            description: description.to_string(),
            current: 0,
            total,
            width: 40,
        }
    }

    pub fn advance(&mut self, amount: usize) {
        self.current = (self.current + amount).min(self.total);
    }

    pub fn set(&mut self, value: usize) {
        self.current = value.min(self.total);
    }

    pub fn render(&self) -> String {
        let pct = if self.total > 0 {
            self.current as f64 / self.total as f64
        } else {
            1.0
        };
        let filled = (self.width as f64 * pct) as usize;
        let empty = self.width.saturating_sub(filled);
        format!(
            "{} [{}{}] {}/{}",
            self.description,
            "=".repeat(filled),
            " ".repeat(empty),
            self.current,
            self.total,
        )
    }

    pub fn is_finished(&self) -> bool {
        self.current >= self.total
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn color_codes() {
        assert_eq!(Color::Red.code(), "\x1b[31m");
        assert_eq!(Color::Reset.code(), "\x1b[0m");
    }

    #[test]
    fn get_terminal_width_at_least_80() {
        assert!(get_terminal_width() >= 80);
    }

    #[test]
    fn get_terminal_height_at_least_24() {
        assert!(get_terminal_height() >= 24);
    }

    #[test]
    fn progress_bar_advance_and_set() {
        let mut pb = ProgressBar::new("test", 100);
        assert_eq!(pb.current, 0);
        pb.advance(10);
        assert_eq!(pb.current, 10);
        pb.set(50);
        assert_eq!(pb.current, 50);
        assert!(!pb.is_finished());
        pb.set(100);
        assert!(pb.is_finished());
    }

    #[test]
    fn all_color_codes() {
        assert_eq!(Color::Reset.code(), "\x1b[0m");
        assert_eq!(Color::Black.code(), "\x1b[30m");
        assert_eq!(Color::Red.code(), "\x1b[31m");
        assert_eq!(Color::Green.code(), "\x1b[32m");
        assert_eq!(Color::Yellow.code(), "\x1b[33m");
        assert_eq!(Color::Blue.code(), "\x1b[34m");
        assert_eq!(Color::Magenta.code(), "\x1b[35m");
        assert_eq!(Color::Cyan.code(), "\x1b[36m");
        assert_eq!(Color::White.code(), "\x1b[37m");
        assert_eq!(Color::BrightBlack.code(), "\x1b[90m");
        assert_eq!(Color::BrightRed.code(), "\x1b[91m");
        assert_eq!(Color::BrightGreen.code(), "\x1b[92m");
        assert_eq!(Color::BrightYellow.code(), "\x1b[93m");
        assert_eq!(Color::BrightBlue.code(), "\x1b[94m");
        assert_eq!(Color::BrightMagenta.code(), "\x1b[95m");
        assert_eq!(Color::BrightCyan.code(), "\x1b[96m");
        assert_eq!(Color::BrightWhite.code(), "\x1b[97m");
    }

    #[test]
    fn progress_bar_render_states() {
        let pb = ProgressBar::new("test", 100);
        let rendered = pb.render();
        assert!(rendered.starts_with("test ["));
        assert!(rendered.contains("0/100"));

        let pb = ProgressBar::new("test", 0);
        let rendered = pb.render();
        assert!(rendered.contains("0/0"));

        let mut pb = ProgressBar::new("test", 100);
        pb.set(100);
        let rendered = pb.render();
        assert!(rendered.contains("100/100"));
    }

    #[test]
    fn confirm_default_yes_with_empty_input() {
        let mut input = b"\n" as &[u8];
        let mut output = Vec::new();
        let result = confirm_with("Proceed?", true, &mut input, &mut output);
        assert!(result.unwrap());
        let out = String::from_utf8(output).unwrap();
        assert!(out.contains("Proceed?"));
        assert!(out.contains("[Y/n]"));
    }

    #[test]
    fn confirm_accepts_yes() {
        let mut input = b"y\n" as &[u8];
        let mut output = Vec::new();
        let result = confirm_with("Go?", false, &mut input, &mut output);
        assert!(result.unwrap());
    }

    #[test]
    fn confirm_accepts_no() {
        let mut input = b"n\n" as &[u8];
        let mut output = Vec::new();
        let result = confirm_with("Go?", true, &mut input, &mut output);
        assert!(!result.unwrap());
    }

    #[test]
    fn confirm_default_no_with_invalid_input() {
        let mut input = b"maybe\n" as &[u8];
        let mut output = Vec::new();
        let result = confirm_with("Go?", false, &mut input, &mut output);
        assert!(!result.unwrap());
    }

    #[test]
    fn confirm_default_yes_with_invalid_input() {
        let mut input = b"maybe\n" as &[u8];
        let mut output = Vec::new();
        let result = confirm_with("Go?", true, &mut input, &mut output);
        assert!(result.unwrap());
    }

    #[test]
    fn ask_returns_default_on_empty() {
        let mut input = b"\n" as &[u8];
        let mut output = Vec::new();
        let result = ask_with("Name", "default_val", &mut input, &mut output);
        assert_eq!(result.unwrap(), "default_val");
    }

    #[test]
    fn ask_returns_input() {
        let mut input = b"custom\n" as &[u8];
        let mut output = Vec::new();
        let result = ask_with("Name", "default_val", &mut input, &mut output);
        assert_eq!(result.unwrap(), "custom");
    }

    #[test]
    fn ask_no_default_shows_no_brackets() {
        let mut input = b"val\n" as &[u8];
        let mut output = Vec::new();
        let result = ask_with("Name", "", &mut input, &mut output);
        assert_eq!(result.unwrap(), "val");
        let out = String::from_utf8(output).unwrap();
        assert!(!out.contains('['));
    }

    #[test]
    fn ask_int_returns_default_on_empty() {
        let mut input = b"\n" as &[u8];
        let mut output = Vec::new();
        let result = ask_int_with("Count", Some(42), &mut input, &mut output);
        assert_eq!(result.unwrap(), 42);
    }

    #[test]
    fn ask_int_returns_input() {
        let mut input = b"99\n" as &[u8];
        let mut output = Vec::new();
        let result = ask_int_with("Count", Some(42), &mut input, &mut output);
        assert_eq!(result.unwrap(), 99);
    }

    #[test]
    fn ask_int_rejects_invalid() {
        let mut input = b"not_a_number\n" as &[u8];
        let mut output = Vec::new();
        let result = ask_int_with("Count", Some(42), &mut input, &mut output);
        assert!(result.is_err());
    }

    #[test]
    fn ask_int_empty_no_default_is_error() {
        let mut input = b"\n" as &[u8];
        let mut output = Vec::new();
        let result = ask_int_with("Count", None, &mut input, &mut output);
        assert!(result.is_err());
    }

    #[test]
    fn is_terminal_returns_bool() {
        let result = is_terminal();
        assert!(result == true || result == false);
    }

    #[test]
    fn progress_bar_set_clamps_at_total() {
        let mut pb = ProgressBar::new("test", 50);
        pb.set(100);
        assert_eq!(pb.current, 50);
        assert!(pb.is_finished());
    }

    #[test]
    fn progress_bar_advance_clamps_at_total() {
        let mut pb = ProgressBar::new("test", 50);
        pb.advance(100);
        assert_eq!(pb.current, 50);
    }
}
